File: cibcc/Go/step17-19/sect1/codegen.go
package main

import (
	"fmt"
)

var labelseq = 0
var argreg = [6]string{"rdi", "rsi", "rdx", "rcx", "r8", "r9"}
var funcname []rune

func gen(node *Node) {
	if node.kind == ND_NUM {
		fmt.Printf("  push %d\n", node.val)
		return
	}
	if node.kind == ND_EXPR_STMT {
		gen(node.lhs)
		fmt.Println("  add rsp, 8")
		return
	}
	if node.kind == ND_VAR {
		gen_addr(node)
		load()
		return
	}
	if node.kind == ND_ASSIGN {
		gen_addr(node.lhs)
		gen(node.rhs)
		store()
		return
	}
	if node.kind == ND_ADDR {
		gen_addr(node.lhs)
		return
	}
	if node.kind == ND_DEREF {
		gen(node.lhs)
		load()
		return
	}
	if node.kind == ND_IF {
		seq := labelseq
		labelseq++
		if node.els != nil {
			gen(node.cond)
			fmt.Println("  pop rax")
			fmt.Println("  cmp rax, 0")
			fmt.Printf("  je .Lelse%d\n", seq)
			gen(node.then)
			fmt.Printf("  jmp .Lend%d\n", seq)
			fmt.Printf(".Lelse%d:\n", seq)
			gen(node.els)
			fmt.Printf(".Lend%d:\n", seq)
		} else {
			gen(node.cond)
			fmt.Println("  pop rax")
			fmt.Println("  cmp rax, 0")
			fmt.Printf("  je .Lend%d\n", seq)
			gen(node.then)
			fmt.Printf(".Lend%d:\n", seq)
		}
		return
	}
	if node.kind == ND_WHILE {
		seq := labelseq
		labelseq++
		fmt.Printf(".Lbegin%d:\n", seq)
		gen(node.cond)
		fmt.Println("  pop rax")
		fmt.Println("  cmp rax, 0")
		fmt.Printf("  je .Lend%d\n", seq)
		gen(node.then)
		fmt.Printf("  jmp .Lbegin%d\n", seq)
		fmt.Printf(".Lend%d:\n", seq)
		return
	}
	if node.kind == ND_FOR {
		seq := labelseq
		labelseq++
		if node.init != nil {
			gen(node.init)
		}
		fmt.Printf(".Lbegin%d:\n", seq)
		if node.cond != nil {
			gen(node.cond)
			fmt.Println("  pop rax")
			fmt.Println("  cmp rax, 0")
			fmt.Printf("  je .Lend%d\n", seq)
		}
		gen(node.then)
		if node.inc != nil {
			gen(node.inc)
		}
		fmt.Printf("  jmp .Lbegin%d\n", seq)
		fmt.Printf(".Lend%d:\n", seq)
		return
	}
	if node.kind == ND_BLOCK {
		for n := node.body; n != nil; n = n.next {
			gen(n)
		}
		return
	}
	if node.kind == ND_FUNCALL {
		nargs := 0
		for n := node.args; n != nil; n = n.next {
			gen(n)
			nargs++
		}
		for i := nargs-1; i >= 0;i-- {
			fmt.Printf("  pop %s\n", argreg[i])
		}
		seq := labelseq
		labelseq++
		fmt.Printf("  mov rax, rsp\n")
		fmt.Printf("  and rax, 15\n")
		fmt.Printf("  jnz .Lcall%d\n", seq)
		fmt.Printf("  mov rax, 0\n")
		fmt.Printf("  call %s\n", string(node.funcname))
		fmt.Printf("  jmp .Lend%d\n", seq)
		fmt.Printf(".Lcall%d:\n", seq)
		fmt.Printf("  sub rsp, 8\n")
		fmt.Printf("  mov rax, 0\n")
		fmt.Printf("  call %s\n", string(node.funcname))
		fmt.Printf("  add rsp, 8\n")
		fmt.Printf(".Lend%d:\n", seq)
		fmt.Printf("  push rax\n");
		return
	}
	if node.kind == ND_RETURN {
		gen(node.lhs)
		fmt.Println("  pop rax")
		fmt.Printf("  jmp .Lreturn.%s\n", string(funcname))
		return
	}

	gen(node.lhs)
	gen(node.rhs)

	fmt.Println("  pop rdi")
	fmt.Println("  pop rax")

	switch node.kind {
		case ND_ADD:
			if node.ty.kind == TY_PTR {
				fmt.Println("  imul rdi, 8")
			}
			fmt.Println("  add rax, rdi")
		case ND_SUB:
			if node.ty.kind == TY_PTR {
				fmt.Println("  imul rdi, 8")
			}
			fmt.Println("  sub rax, rdi")
		case ND_MUL:
			fmt.Println("  imul rax, rdi")
		case ND_DIV:
			fmt.Println("  cqo")
			fmt.Println("  idiv rdi")
		case ND_EQ:
			fmt.Println("  cmp rax, rdi")
			fmt.Println("  sete al");
			fmt.Println("  movzb rax, al")
		case ND_NE:
			fmt.Println("  cmp rax, rdi")
			fmt.Println("  setne al");
			fmt.Println("  movzb rax, al")
		case ND_LT:
			fmt.Println("  cmp rax, rdi")
			fmt.Println("  setl al");
			fmt.Println("  movzb rax, al")
		case ND_LE:
			fmt.Println("  cmp rax, rdi")
			fmt.Println("  setle al");
			fmt.Println("  movzb rax, al")
	}

	fmt.Println("  push rax")
}

func codegen(prog *Function) {
	fmt.Println(".intel_syntax noprefix")

	for fn := prog; fn != nil; fn = fn.next {
		fmt.Printf(".global %s\n", string(fn.name))
		fmt.Printf("%s:\n", string(fn.name))
		funcname = fn.name

		fmt.Println("  push rbp")
		fmt.Println("  mov rbp, rsp")
		fmt.Printf("  sub rsp, %d\n", fn.stack_size)

		i := 0
		for vl := fn.params; vl != nil; vl = vl.next {
			va := vl.va
			fmt.Printf("  mov [rbp-%d], %s\n", va.offset, argreg[i])
			i++
		}

		for n := fn.node; n != nil; n = n.next {
			gen(n)
		}

		fmt.Printf(".Lreturn.%s:\n", string(funcname))
		fmt.Println("  mov rsp, rbp")
		fmt.Println("  pop rbp")
		fmt.Println("  ret")
	}
}

func gen_addr(node *Node) {
	if node.kind == ND_VAR {
		fmt.Printf("  lea rax, [rbp-%d]\n", node.va.offset)
		fmt.Println("  push rax")
		return
	}
	if node.kind == ND_DEREF {
		gen(node.lhs)
		return
	}
	fmt.Println("not an value")
}

func load() {
	fmt.Println("  pop rax")
	fmt.Println("  mov rax, [rax]")
	fmt.Println("  push rax")
}

func store() {
	fmt.Println("  pop rdi")
	fmt.Println("  pop rax")
	fmt.Println("  mov [rax], rdi")
	fmt.Println("  push rdi")
}

File: cibcc/java/step17-19/sect1/Codegen.java
public class Codegen {

	int labelseq = 0;
	String argreg[] = {"rdi", "rsi", "rdx", "rcx", "r8", "r9"};
	String funcname;

	public void codegen(Function prog) {
		System.out.println(".intel_syntax noprefix");

		for (Function fn = prog; fn != null; fn = fn.next) {
			System.out.printf(".global %s\n", fn.name);
			System.out.printf("%s:\n", fn.name);
			funcname = fn.name;

			System.out.println("  push rbp");
			System.out.println("  mov rbp, rsp");
			System.out.printf("  sub rsp, %d\n", fn.stack_size);

			int i = 0;
			for (VarList vl = fn.params; vl != null; vl = vl.next) {
				Var va = vl.va;
				System.out.printf("  mov [rbp-%d], %s\n", va.offset, argreg[i++]);
			}
			for (Node n = fn.node; n != null; n = n.next) {
				gen(n);
			}

			System.out.printf(".Lreturn.%s:\n", funcname);
			System.out.println("  mov rsp, rbp");
			System.out.println("  pop rbp");
			System.out.println("  ret");
		}
	}

	private void gen(Node node) {

		if (node.kind == NodeKind.ND_NUM) {
			System.out.printf("  push %d\n", node.val);
			return;
		}
		if (node.kind == NodeKind.ND_EXPR_STMT) {
			gen(node.lhs);
			System.out.println("  add rsp, 8");
			return;
		}
		if (node.kind == NodeKind.ND_VAR) {
			gen_addr(node);
			load();
			return;
		}
		if (node.kind == NodeKind.ND_ASSIGN) {
			gen_addr(node.lhs);
			gen(node.rhs);
			store();
			return;
		}
		if (node.kind == NodeKind.ND_ADDR) {
			gen_addr(node.lhs);
			return;
		}
		if (node.kind == NodeKind.ND_DEREF) {
			gen(node.lhs);
			load();
			return;
		}
		if (node.kind == NodeKind.ND_IF) {
			int seq = labelseq;
			labelseq++;
			if (node.els !=null) {
				gen(node.cond);
				System.out.println("  pop rax");
				System.out.println("  cmp rax, 0");
				System.out.printf("  je .Lelse%d\n", seq);
				gen(node.then);
				System.out.printf(".Lelse%d:\n", seq);
				gen(node.els);
			} else {
				gen(node.cond);
				System.out.println("  pop rax");
				System.out.println("  cmp rax, 0");
				System.out.printf("  je .Lend%d\n", seq);
				gen(node.then);
				System.out.printf(".Lend%d:\n", seq);
			}
			return;
		}
		if (node.kind == NodeKind.ND_WHILE) {
			int seq = labelseq++;
			System.out.printf(".Lbegin%d:\n", seq);
			gen(node.cond);
			System.out.println("  pop rax");
			System.out.println("  cmp rax, 0");
			System.out.printf("  je .Lend%d\n", seq);
			gen(node.then);
			System.out.printf("  jmp .Lbegin%d\n", seq);
			System.out.printf(".Lend%d:\n", seq);
			return;
		}
		if (node.kind == NodeKind.ND_FOR) {
			int seq = labelseq++;
			if (node.init != null) {
				gen(node.init);
			}
			System.out.printf(".Lbegin%d:\n", seq);
			if (node.cond != null) {
				gen(node.cond);
				System.out.println("  pop rax");
				System.out.println("  cmp rax, 0");
				System.out.printf("  je .Lend%d\n", seq);
			}
			gen(node.then);
			if (node.inc != null) {
				gen(node.inc);
			}
			System.out.printf("  jmp .Lbegin%d\n", seq);
			System.out.printf(".Lend%d:\n", seq);
			return;
		}
		if (node.kind == NodeKind.ND_BLOCK) {
			for (Node n = node.body; n != null; n = n.next) {
				gen(n);
			}
			return;
		}
		if (node.kind == NodeKind.ND_FUNCALL) {
			int nargs = 0;
			for (Node arg = node.args; arg != null; arg = arg.next) {
				gen(arg);
				nargs++;
			}
			for (int i = nargs - 1; i >= 0; i--) {
				System.out.printf("  pop %s\n", argreg[i]);
			}
			int seq = labelseq++;
			System.out.println("  mov rax, rsp");
			System.out.println("  and rax, 15");
			System.out.printf("  jnz .Lcall%d\n", seq);
			System.out.println("  mov rax, 0");
			System.out.printf("  call %s\n", node.funcname);
			System.out.printf("  jmp .Lend%d\n", seq);
			System.out.printf(".Lcall%d:\n", seq);
			System.out.println("  sub rsp, 8");
			System.out.println("  mov rax, 0");
			System.out.printf("  call %s\n", node.funcname);
			System.out.println("  add rsp, 8");
			System.out.printf(".Lend%d:\n", seq);
			System.out.println("  push rax");
			return;
		}
		if (node.kind == NodeKind.ND_RETURN) {
			gen(node.lhs);
			System.out.println("  pop rax");
			System.out.printf("  jmp .Lreturn.%s\n", funcname);
			return;
		}

		gen(node.lhs);
		gen(node.rhs);

		System.out.println("  pop rdi");
		System.out.println("  pop rax");

		switch (node.kind) {
			case ND_ADD:
				if (node.ty.kind.equals(Type.TypeKind.TY_PTR)) {
					System.out.println("  imul rdi, 8");
				} else if (node.ty == null) {
					System.err.println("Error in Codegen: Node has no Type info. NodeKind: " + node.kind);
				}
				System.out.println("  add rax, rdi");
				break;
			case ND_SUB:
				if (node.ty.kind.equals(Type.TypeKind.TY_PTR)) {
					System.out.println("  imul rdi, 8");
				}
				System.out.println("  sub rax, rdi");
				break;
			case ND_MUL:
				System.out.println("  imul rax, rdi");
				break;
			case ND_DIV:
				System.out.println("  cqo");
				System.out.println("  idiv rdi");
				break;
			case ND_EQ:
				System.out.println("  cmp rax, rdi");
				System.out.println("  sete al");
				System.out.println("  movzb rax, al");
				break;
			case ND_NE:
				System.out.println("  cmp rax, rdi");
				System.out.println("  setne al");
				System.out.println("  movzb rax, al");
				break;
			case ND_LT:
				System.out.println("  cmp rax, rdi");
				System.out.println("  setl al");
				System.out.println("  movzb rax, al");
				break;
			case ND_LE:
				System.out.println("  cmp rax, rdi");
				System.out.println("  setle al");
				System.out.println("  movzb rax, al");
				break;
		}
		System.out.println("  push rax");
	}

	private void gen_addr(Node node) {
		if (node.kind == NodeKind.ND_VAR) {
			System.out.printf("  lea rax, [rbp-%d]\n", node.va.offset);
			System.out.println("  push rax");
			return;
		}
		if (node.kind == NodeKind.ND_DEREF) {
			gen(node.lhs);
			return;
		}
		System.out.println("not an lvalue");
	}

	private void load() {
		System.out.println("  pop rax");
		System.out.println("  mov rax, [rax]");
		System.out.println("  push rax");
	}

	private void store() {
		System.out.println("  pop rdi");
		System.out.println("  pop rax");
		System.out.println("  mov [rax], rdi");
		System.out.println("  push rdi");
	}
}

	
